packages/langium/test/grammar/langium-grammar-semantic-token-provider.test.ts

```diff
@@ -0,0 +1,108 @@
+/******************************************************************************
+ * Copyright 2022 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+import { highlightHelper, expectSemanticToken } from '../../src/test/langium-test';
+import { createLangiumGrammarServices } from '../../src';
+import { SemanticTokenTypes } from 'vscode-languageserver';
+
+const services = createLangiumGrammarServices();
+const highlight = highlightHelper(services.grammar);
+
+describe('Langium grammar semantic token provider', () => {
+    test('should highlight primitive and referenced types as token type "type"', async () => {
+        // arrange
+        const grammarText = `grammar Test
+        interface X {
+            x: <|string|> | <|AbstractDefinition|>
+        }
+        `.trim();
+
+        // act
+        const tokens = await highlight(grammarText);
+
+        // assert
+        expectSemanticToken(tokens, {
+            rangeIndex: 0,
+            tokenType: SemanticTokenTypes.type,
+        });
+        expectSemanticToken(tokens, {
+            rangeIndex: 1,
+            tokenType: SemanticTokenTypes.type,
+        });
+    });
+
+    test('should highlight assignment\'s feature as token type "property"', async () => {
+        // arrange
+        const grammarText = `grammar Test
+        entry Main: <|name|>=ID;
+        terminal ID: /[a-z]+/;
+        `.trim();
+
+        // act
+        const tokens = await highlight(grammarText);
+
+        // assert
+        expectSemanticToken(tokens, {
+            tokenType: SemanticTokenTypes.property,
+        });
+    });
+
+    test('should highlight action\'s feature as token type "property"', async () => {
+        // arrange
+        const grammarText = `grammar Test
+        interface A {name: string; main: Main;}
+        entry Main: {infer A.<|main|>=current} name=ID;
+        terminal ID: /[a-z]+/;
+        `.trim();
+
+        // act
+        const tokens = await highlight(grammarText);
+
+        // assert
+        expectSemanticToken(tokens, {
+            tokenType: SemanticTokenTypes.property,
+        });
+    });
+
+    test.skip('should highlight return type\'s name as token type "type"', async () => {
```

Comment:
This test is suspicious. It fails because it never visits a return type.

Reply:
You mean that the highlighting doesn't work as expected when writing `X returns Y: ...;`? I would think that this fix would be part of this PR.

Reply:
No, it works. I had the wrong assumptions in my mind. Return type is defined differently for terminal and parser rules.